// Copyright 2019 Zeppelin Bend Pty Ltd
// This file is part of cimbend.
// 
// cimbend is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// cimbend is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
// 
// You should have received a copy of the GNU Affero General Public License
// along with cimbend.  If not, see <https://www.gnu.org/licenses/>.


syntax = "proto3";
option java_multiple_files = true;
option java_package = "com.zepben.protobuf.cim.iec61970.base.wires";
package zepben.cim.iec61970.base.wires;

import "zepben/cim/iec61970/base/diagramlayout/DiagramObject.proto";
import "zepben/cim/iec61970/base/core/Terminal.proto";
import "zepben/cim/iec61968/common/Location.proto";
import "zepben/cim/iec61970/base/meas/Control.proto";
import "zepben/cim/iec61970/base/meas/Measurement.proto";
import "zepben/cim/iec61970/base/wires/RatioTapChanger.proto";
import "zepben/cim/iec61970/base/wires/WindingConnection.proto";

/**
 * A PowerTransformerEnd is associated with each Terminal of a PowerTransformer.
 * The impedance values r, r0, x, and x0 of a PowerTransformerEnd represents a star equivalent as follows
 * <p>
 * 1) for a two Terminal PowerTransformer the high voltage PowerTransformerEnd has non zero values on r, r0, x, and x0
 *    while the low voltage PowerTransformerEnd has zero values for r, r0, x, and x0.
 * 2) for a three Terminal PowerTransformer the three PowerTransformerEnds represents a star equivalent with each leg in
 *    the star represented by r, r0, x, and x0 values.
 * 3) for a PowerTransformer with more than three Terminals the PowerTransformerEnd impedance values cannot be used.
 *    Instead use the TransformerMeshImpedance or split the transformer into multiple PowerTransformers.
 */
message PowerTransformerEnd {

    /**
     * Master resource identifier issued by a model authority. The mRID is unique within an exchange context. Global uniqueness
     * is easily achieved by using a UUID,  as specified in RFC 4122, for the mRID. The use of UUID is strongly recommended.
     * <p>
     * For CIMXML data files in RDF syntax conforming to IEC 61970-552 Edition 1, the mRID is mapped to rdf:ID or rdf:about attributes
     * that identify CIM object elements.
     */
    string mRID = 1;

    /**
     * The name is any free human readable and possibly non unique text naming the object.
     */
    string name = 2;

    /**
     * If true, the equipment is in service.
     */
    bool inService = 3;

    /**
     * If true, the equipment is normally in service.
     */
    bool normallyInService = 4;

    /**
     * Base voltage of this conducting equipment.  Use only when there is no voltage level container used and only one base
     * voltage applies.  For example, not used for transformers.
     */
    string baseVoltageMRID = 5;

    /**
     * Conducting equipment have terminals that may be connected to other conducting equipment terminals via connectivity nodes
     * or topological nodes.
     */
    repeated iec61970.base.core.Terminal terminals = 6;

    /**
     * Normal apparent power rating. The attribute shall be a positive value. For a two-winding transformer the values
     * for the high and low voltage sides shall be identical.
     */
    double ratedS = 7;

    /**
     * Rated voltage: phase-phase for three-phase windings, and either phase-phase or phase-neutral for single-phase windings.
     * A high voltage side, as given by TransformerEnd.endNumber, shall have a ratedU that is greater or equal than
     * ratedU for the lower voltage sides
     */
    double ratedU = 8;

    /**
     * Resistance (star-model) of the transformer end.
     * The attribute shall be equal or greater than zero for non-equivalent transformers.
     */
    double r = 9;

    /**
     * Positive sequence series reactance (star-model) of the transformer end.
     */
    double x = 10;

    /**
     * Zero sequence series resistance (star-model) of the transformer end.
     */
    double r0 = 11;

    /**
     * Zero sequence series reactance of the transformer end.
     */
    double x0 = 12;

    /**
     * Kind of connection.
     */
    zepben.cim.iec61970.base.wires.winding.WindingConnection connectionKind = 13;

    /**
     * The tap changer associated with this PowerTransformerEnd
     */
    zepben.cim.iec61970.base.wires.RatioTapChanger ratioTapChanger = 14;

    /**
     * The diagram objects that are associated with the domain object.
     * <p>
     * The diagram objects are returned in a map sorted by the diagram they live in.
     */
    repeated iec61970.base.diagramlayout.DiagramObject diagramObjects = 16;

    /**
     * Datasheet information for this power system resource.
     */
    string assetInfoMRID = 17;

    /**
     * Location of this power system resource.
     */
    iec61968.common.Location location = 18;

}
